### PR TITLE
support serialize/deserialize instance of custom class

### PR DESCRIPTION
**Pull request by zenghsh3**

### CODE REVIEW

**Comment by zenghsh3** [comment]
Result in 12 CPU machine (Intel(R) Xeon(R) Gold 6128 CPU @ 3.40GHz), Ubuntu:

Case 0, Average dump and load argument time: 2.1354129314422607
Case 1, Average dump and load argument time: 0.012008333206176757
Case 2, Average dump and load argument time: 0.5284070491790771

Case 0, Average dump and load return time: 2.165187621116638
Case 1, Average dump and load return time: 0.011833405494689942
Case 2, Average dump and load return time: 0.5375057220458984

The speed is almost same with baseline #76

**Comment by TomorrowIsAnOtherDay** [comment]
> Result in 12 CPU machine (Intel(R) Xeon(R) Gold 6128 CPU @ 3.40GHz), Ubuntu:
> 
> Case 0, Average dump and load argument time: 2.1354129314422607
> Case 1, Average dump and load argument time: 0.012008333206176757
> Case 2, Average dump and load argument time: 0.5284070491790771
> 
> Case 0, Average dump and load return time: 2.165187621116638
> Case 1, Average dump and load return time: 0.011833405494689942
> Case 2, Average dump and load return time: 0.5375057220458984
> 
> The speed is almost same with baseline #76

Great!
I have added a tag(`performance_test_needed`) for this issue so that we can find previous performance  record easily.

**Comment by TomorrowIsAnOtherDay** [comment]
Since we are using a global object `serialization_context` for serialization and  deserialization, unitest for thread-safe test is a must. Could you add unit test for that?

**Comment by zenghsh3** [comment]
Latest commit speed:

Case 0, Average dump and load argument time: 2.1217625617980955
Case 1, Average dump and load argument time: 0.0116349458694458
Case 2, Average dump and load argument time: 0.5327484369277954

Case 0, Average dump and load return time: 2.085127091407776
Case 1, Average dump and load return time: 0.011567068099975587
Case 2, Average dump and load return time: 0.5333646059036254

**Comment by zenghsh3** [comment]
Test throughput with IMPALA algorithm in a P40 GPU machine with 28 CPUs.
Sample steps in 10 minutes (28 actors):

| Version     | Sample steps |
|-------------|--------------|
| parl==1.1.0 | 4410500      |
| This PR     | 4427000      |

So there should be no big difference in throughput.

**Comment by TomorrowIsAnOtherDay** [comment]
Have you added unit test code for thread-safe test?

**Comment by zenghsh3** [comment]
Is this commit enough?https://github.com/PaddlePaddle/PARL/pull/77/commits/056604628effcf74e954fe0e991a083ba9c9d5d2

**Comment by TomorrowIsAnOtherDay** [comment]
No. As we have discussed before, we can test whether the code is thread-safe by launching 50 threads, with each thread counting from 1 to 10000. When all the threads are finished, their result must be equal to 10000.

File: parl/utils/exceptions.py [hunk 1]
```diff
@@ -28,6 +28,9 @@ class SerializeError(UtilsError):
     """
 
     def __init__(self, error_info):
+        error_info = (
+            'Serialize error, you may have provided a pyarrow ' +
```
Comment (TomorrowIsAnOtherDay) [inline parl/utils/exceptions.py]: you may have provided an object that cannot be serialized by pyarrow.

File: parl/utils/exceptions.py [hunk 2]
```diff
@@ -40,6 +43,10 @@ class DeserializeError(UtilsError):
     """
 
     def __init__(self, error_info):
+        error_info = (
+            'Deserialize error, you may have provided a pyarrow' +
+            'non-deserializable object. Detailed error:\n{}'.format(error_info)
```
Comment (TomorrowIsAnOtherDay) [inline parl/utils/exceptions.py]: you may have provided an object that cannot be deserialized by pyarrow.

Comment (zenghsh3) [reply parl/utils/exceptions.py]: done

File: parl/utils/exceptions.py (hunk 1 above)
Comment (zenghsh3) [reply parl/utils/exceptions.py]: done

**Review by TomorrowIsAnOtherDay — APPROVED** [review]

**Pull request merged**